### PR TITLE
msg: add min delay packets support for mons

### PR DESCRIPTION
**Pull request by shun-s**
make packets between mons support DSCP (TOS)，just like osd heartbeat

when suffering network bottleneck, like switch qos control for only 50Mbps,
mon cluster may stuck in electing for long time and elects again and again, which intrduces many troubles, like it's hard to find where the problem is. even more worse, it may lead to IO hang when one osd can't serve IO. so it's better to keep mon cluster stable always.

Signed-off-by: Song Shun <song.shun3@zte.com.cn>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by shun-s** [comment]
@tchaikov @idryomov   would you mind taking a review  ? thanks

File: src/common/options.cc [hunk 1]
```diff
@@ -2228,6 +2228,11 @@ std::vector<Option> get_global_options() {
 		     "only be rejected later, when they are used.")
     .set_flag(Option::FLAG_RUNTIME),
 
+    Option("mon_min_delay_socket", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
+    .set_default(false)
+    .set_description("priority packets between mons")
+    .add_service("mon"),
```
Comment (idryomov) [inline src/common/options.cc:2234]: I'd rename this option to `mon_use_min_delay_socket` and add a `see_also` for `osd_heartbeat_use_min_delay_socket`.

File: src/msg/async/AsyncConnection.cc [hunk 2]
```diff
@@ -394,6 +394,12 @@ void AsyncConnection::process() {
 
       SocketOptions opts;
       opts.priority = async_msgr->get_socket_priority();
+      if (async_msgr->cct->_conf->mon_min_delay_socket) {
+          if (async_msgr->get_mytype() == CEPH_ENTITY_TYPE_MON
+              && peer_is_mon()) {
```
Comment (idryomov) [inline src/msg/async/AsyncConnection.cc:399]: `&&` should be on the previous line.

File: src/msg/async/AsyncConnection.cc [hunk 3]
```diff
@@ -443,7 +449,18 @@ void AsyncConnection::process() {
     case STATE_ACCEPTING: {
       center->create_file_event(cs.fd(), EVENT_READABLE, read_handler);
       state = STATE_CONNECTION_ESTABLISHED;
-
+      if (async_msgr->cct->_conf->mon_min_delay_socket) {
+        if (async_msgr->get_mytype() == CEPH_ENTITY_TYPE_MON
+            && peer_is_mon()) {
```
Comment (idryomov) [inline src/msg/async/AsyncConnection.cc:454]: `&&` should be on the previous line.

File: src/msg/async/AsyncConnection.cc [hunk 4]
```diff
@@ -443,7 +449,18 @@ void AsyncConnection::process() {
...
+          ssize_t r = cs.set_priority(cs.fd(), SOCKET_PRIORITY_MIN_DELAY, target_addr.get_family());
```
Comment (idryomov) [inline src/msg/async/AsyncConnection.cc:455]: This line is too long.

File: src/msg/async/PosixStack.cc [hunk 5]
```diff
@@ -206,6 +206,9 @@ class PosixConnectedSocketImpl final : public ConnectedSocketImpl {
   void close() override {
     ::close(_fd);
   }
+  int set_priority(int sd, int prio, int domain) {
+    handler.set_priority(sd, prio, domain);
```
Comment (idryomov) [inline src/msg/async/PosixStack.cc:210]: This is missing a return value.

File: src/msg/async/Stack.h [hunk 6]
```diff
@@ -32,6 +32,7 @@ class ConnectedSocketImpl {
   virtual void shutdown() = 0;
   virtual void close() = 0;
   virtual int fd() const = 0;
+  virtual int set_priority(int sd, int prio, int domain) = 0;
```
Comment (idryomov) [inline src/msg/async/Stack.h:35]: Again, given that `NetHandler::set_priority()` swallows all errors, what is the purpose of returning an int here?

File: src/msg/async/AsyncConnection.cc [hunk 7]
```diff
@@ -443,7 +449,18 @@ void AsyncConnection::process() {
...
+          if (r < 0) {
+            ldout(async_msgr->cct, 0) << __func__ << " couldn't set SO_PRIORITY to "
+                 << SOCKET_PRIORITY_MIN_DELAY << ": " << cpp_strerror(r) << dendl;
+            protocol->fault();
+            return;
+          }
```
Comment (idryomov) [inline src/msg/async/AsyncConnection.cc:461]: Given that `NetHandler::set_priority()` swallows all errors, this looks like a dead branch.

File: src/msg/async/rdma/RDMAStack.h [hunk 8]
```diff
@@ -218,6 +218,7 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
   virtual void shutdown() override;
   virtual void close() override;
   virtual int fd() const override { return notify_fd; }
+  virtual int set_priority(int sd, int prio, int domain) {return 0;}
```
Comment (idryomov) [inline src/msg/async/rdma/RDMAStack.h:221]: Not sure about DPDK, but why are you stubbing out RDMA?  The existing messenger-wide setting seems to cover RDMA, see `RDMAConnectedSocketImpl::try_connect()` and `RDMAServerSocketImpl::accept()`.

**Review by idryomov — CHANGES_REQUESTED** [review]

File: src/msg/async/AsyncConnection.cc (hunk 7 above)
Comment (shun-s) [reply src/msg/async/AsyncConnection.cc:461]: returning error and retry  may be better than swallowing all errors, i will change cs.set_priority code to return error .

File: src/msg/async/Stack.h (hunk 6 above)
Comment (shun-s) [reply src/msg/async/Stack.h:35]: change to not swallows all errors, thanks

File: src/msg/async/rdma/RDMAStack.h (hunk 8 above)
Comment (shun-s) [reply src/msg/async/rdma/RDMAStack.h:221]: OK,  forget it due to first coding on outdated master

File: src/common/options.cc (hunk 1 above)
Comment (shun-s) [reply src/common/options.cc]: done

File: src/msg/async/AsyncConnection.cc (hunk 2 above)
Comment (shun-s) [reply src/msg/async/AsyncConnection.cc]: OK

File: src/msg/async/AsyncConnection.cc (hunk 3 above)
Comment (shun-s) [reply src/msg/async/AsyncConnection.cc]: OK

File: src/msg/async/AsyncConnection.cc (hunk 4 above)
Comment (shun-s) [reply src/msg/async/AsyncConnection.cc]: modify done

File: src/msg/async/PosixStack.cc (hunk 5 above)
Comment (shun-s) [reply src/msg/async/PosixStack.cc:210]: i think it's better to change set_priority to void

File: src/msg/async/AsyncConnection.cc (hunk 7 above)
Comment (shun-s) [reply src/msg/async/AsyncConnection.cc]: already remove this branch

File: src/msg/async/rdma/RDMAStack.h (hunk 8 above)
Comment (shun-s) [reply src/msg/async/rdma/RDMAStack.h]: 1. add modifycation for RDMAConnectedSocketImpl
2.  DPDK seems not support  all socket configs yet, so choose to implement always support.

**Comment by shun-s** [comment]
@idryomov  all your concern seems to be fixed, please take a review again. thanks

File: src/msg/async/dpdk/DPDKStack.h [hunk 9]
```diff
@@ -46,6 +46,7 @@ class DPDKServerSocketImpl : public ServerSocketImpl {
   virtual int fd() const override {
     return _listener.fd();
   }
+  virtual void set_priority(int sd, int prio, int domain) {}
```
Comment (idryomov) [inline src/msg/async/dpdk/DPDKStack.h:49]: Missing `override`?

File: src/msg/async/AsyncConnection.cc [hunk 10]
```diff
@@ -443,7 +449,14 @@ void AsyncConnection::process() {
     case STATE_ACCEPTING: {
       center->create_file_event(cs.fd(), EVENT_READABLE, read_handler);
       state = STATE_CONNECTION_ESTABLISHED;
-
+      if (async_msgr->cct->_conf->mon_use_min_delay_socket) {
+        if (async_msgr->get_mytype() == CEPH_ENTITY_TYPE_MON &&
+            peer_is_mon()) {
+          cs.set_priority(cs.fd(),
+                          SOCKET_PRIORITY_MIN_DELAY,
```
Comment (idryomov) [inline src/msg/async/AsyncConnection.cc:456]: Nit: no need to line break between `cs.fd()` and `SOCKET_PRIORITY_MIN_DELAY`.

File: src/msg/async/PosixStack.cc [hunk 11]
```diff
@@ -206,6 +206,9 @@ class PosixConnectedSocketImpl final : public ConnectedSocketImpl {
   void close() override {
     ::close(_fd);
   }
+  void set_priority(int sd, int prio, int domain) {
```
Comment (idryomov) [inline src/msg/async/PosixStack.cc:209]: Missing `override`?

**Review by idryomov — COMMENTED** [review]
A couple minor nits, nothing else jumps out at me.

You don't mention what problem does prioritizing mon <-> mon communication solve and what sort of improvement is expected to be observed with `mon_use_min_delay_socket` turned on though.  It would be great if that was explained in the commit message.

File: src/msg/async/PosixStack.cc (hunk 11 above)
Comment (shun-s) [reply src/msg/async/PosixStack.cc]: done

File: src/msg/async/AsyncConnection.cc (hunk 10 above)
Comment (shun-s) [reply src/msg/async/AsyncConnection.cc]: done

File: src/msg/async/dpdk/DPDKStack.h (hunk 9 above)
Comment (shun-s) [reply src/msg/async/dpdk/DPDKStack.h]: done

**Comment by shun-s** [comment]
https://github.com/ceph/ceph/pull/37767#pullrequestreview-516868292
// please see commit msg

**Comment by idryomov** [comment]
> when suffering network bottleneck, like switch qos control for only 50Mbps,
> mon cluster may stuck in electing for long time and elects again and again, which intrduces many troubles, like it's hard to find where the problem is. even more worse, it may lead to IO hang when one osd can't serve IO. so it's better to keep mon cluster stable always.

Did you actually see it happen on one of your clusters?

Was this patch alone sufficient or did you have to tweak other settings (whether on the ceph side or on the networking side)?

**Review by idryomov — COMMENTED** [review]
The commit message should be line wrapped at 72 characters.  And there is a typo: "intrduces".

**Comment by shun-s** [comment]
>Did you actually see it happen on one of your clusters?
>Was this patch alone sufficient or did you have to tweak other settings (whether on the ceph side or on the networking side)?
exactly happend in our cluster during test  and when setting mon_use_min_delay_socket to true, mons keep stable most time.

**Comment by shun-s** [comment]
> The commit message should be line wrapped at 72 characters. And there is a typo: "intrduces"

done, please review again

**Review by idryomov — APPROVED** [review]

**Comment by neha-ojha** [comment]
adding @gregsfortytwo as reviewer since he reviewed https://github.com/ceph/ceph/pull/3353 too

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-11-14_09:04:49-rados-wip-kefu-testing-2020-11-14-1525-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-11-15_03:36:16-rados-wip-kefu-testing-2020-11-14-1526-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-11-15_12:08:30-rados-wip-kefu-testing-2020-11-14-1526-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-11-15_16:16:56-rados-wip-kefu-testing-2020-11-14-1526-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/48220
- https://tracker.ceph.com/issues/48030

**Comment by tchaikov** [comment]
@shun-s needs rebase

**Comment by tchaikov** [comment]
@shun-s ping?

**Comment by shun-s** [comment]
@tchaikov  sorry for late reply,  already rebase, please check again. many thanks

**Comment by shun-s** [comment]
@tchaikov ping

**Comment by idryomov** [comment]
The resolution is wrong.  Please rebase on latest master -- note that config options are now defined in yaml files.

**Comment by rzarzynski** [comment]
@shun-s: ping.

**Comment by tchaikov** [comment]
changelog

- rebase against main HEAD
- use the new yaml-based option generation machinery

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Pull request merged**